Add ability to split integration tests into different groups

blakerouse opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

Adds a required field `Group` to `define.Requirements`. Adds ability to select a specific group of tests with `TEST_GROUPS` environment variable. Groups batches by `Group` name splitting the testing load across multiple instances.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Allows tests to be grouped to different instances. Allowing more tests to run concurrently to speed up the time it takes for integration tests to run. Allows the selection of running only a group of tests.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- [x] I have added an integration test or an E2E test

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

Run `mage integration:test` see that multiple instances are spawned instead of only one per platform OS/architecture.

Select a specific group with `TEST_GROUPS="default" mage integration:test`.

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3544/detail/PR-3544/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3544/detail/PR-3544/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3544/detail/PR-3544/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3544/detail/PR-3544/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-agent_3544.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-10-05T15:46:49.479Z&rangeTo=2023-10-05T16:06:49.479Z&transactionName=BUILD+elastic-agent%2Felastic-agent-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=dc215138f6689e1b127833dfa67488b3&transactionId=0af8eb638dc68732)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-10-05T15:56:49.479+0000


* Duration: 25 min 13 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 6485  |
| Skipped      | 59 |
| Total        | 6544   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages.


- `run integration tests` : Run the Elastic Agent Integration tests.


- `run end-to-end tests` : Generate the packages and run the E2E Tests.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 98.78% (`81/82`) | :+1: 
Files | 67.003% (`199/297`) | :+1: 
Classes | 65.642% (`363/553`) | :+1: 
Methods | 52.93% (`1147/2167`) | :-1: -0.024
Lines | 38.457% (`13077/34004`) | :+1: 0.005
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

cmacknz reviewed (commented):
I like this and it the changes are surprisingly simple. Thanks!

I think the hardest thing here will be making sure that it is well understood and used properly. At minimum we will need some explanation of this in https://github.com/elastic/elastic-agent/blob/main/docs/test-framework-dev-guide.md as well.

I have two thoughts to try to help this:

1. Consider making a `ShardID` mandatory and predefine a default shard ID for all tests to reuse. This is more likely to work properly if someone copy/pastes from an existing test IMO, because people will see there is always a shard ID and that some tests use a default. Without this it is harder to notice that ShardID is sometimes undefined.


2. We should call this something other than `ShardID`. Giving framework users control over which tests run where makes sense, but the concept of sharding doesn't have a default definition for test frameworks. This also avoids requiring users to understand what sharding means in general. As for what else to call it, maybe `TestMachineGroup` to try to communicate that all tests in the same group will run on the same VM.

blakerouse commented:
@cmacknz I took your feedback and change it to `Group` as in its a group of tests. I made `Group` a required field in `define.Requirements` and then divided the tests up into groups. I also added `TEST_GROUPS` environment variable to allow the selection of only running a group of tests. I added to the integration framework developer guide as well, explaining groups and how to select specific groups with `TEST_GROUPS`.

cmacknz commented on docs/test-framework-dev-guide.md:
```diff
@@ -67,6 +67,16 @@ between, and it can be very specific or not very specific.
 
+#### Selecting specific group
+
+By default, the running will run all test groups. When working on groups of tests it's better to limit to a specific
```
```suggestion
By default, the runner will run all test groups. Each group runs on a dedicated machine instance. When working on groups of tests it's better to limit to a specific
```

cmacknz commented on pkg/testing/define/batch.go:
```diff
@@ -41,6 +41,10 @@ var defaultOS = []OS{
 type Batch struct {
+	// Group must be set on each test to define which group the tests belongs.
+	// Tests that are in the same group are executed on the same runner.
+	Group string `json:"shard_id,omitempty"`
```
```suggestion
	Group string `json:"group,omitempty"`
```

cmacknz commented on pkg/testing/define/requirements.go:
```diff
@@ -82,6 +82,13 @@ type Stack struct {
 
 // Requirements defines the testing requirements for the test to run.
 type Requirements struct {
+	// Group must be set on each test to define which group the tests belongs.
```
```suggestion
	// Group must be set on each test to define which group the tests belongs to.
```

cmacknz commented on pkg/testing/define/requirements.go:
```diff
@@ -82,6 +82,13 @@ type Stack struct {
+	// Tests that are in the same group are executed on the same runner.
+	//
+	// Useful when a tests take a long time to complete and sharding them across multiple
+	// host can improve the total amount of time to complete all the tests.
```
```suggestion
	// hosts can improve the total amount of time to complete all the tests.
```

cmacknz commented on pkg/testing/define/requirements.go:
```diff
@@ -82,6 +82,13 @@ type Stack struct {
+	// Group must be set on each test to define which group the tests belongs.
+	// Tests that are in the same group are executed on the same runner.
+	//
+	// Useful when a tests take a long time to complete and sharding them across multiple
```
```suggestion
	// Useful when tests take a long time to complete and sharding them across multiple
```

cmacknz commented on docs/test-framework-dev-guide.md:
```diff
@@ -168,6 +178,17 @@ the `github.com/elastic/elastic-agent/pkg/testing/define` package for the test
 
+### Test group
+
+Every `define.Require` must define a `Group` that it belongs too. All tests with in the same group are executed
+on the same instance. Placing similar tests in the same group allows those tests to run on its own instance
```
```suggestion
Every `define.Require` must define a `Group` that it belongs too. Each group is executed on a separate instance with all tests with in the same group executed
on the same instance. Placing similar tests in the same group allows those tests to run on its own instance
```

cmacknz commented on testing/integration/fqdn_test.go:
```diff
@@ -32,6 +32,7 @@ import (
 
 func TestFQDN(t *testing.T) {
 	info := define.Require(t, define.Requirements{
+		Group: "fqdn",
```
Is a test by itself in its own group equal to setting `Isolate: true`? Do we need both ways to control this?

blakerouse replied:
Actually I don't think that we need both. I remove isolate as you can now just place an individual test in its own group.

cmacknz commented on testing/integration/proxy_url_test.go:
```diff
@@ -302,6 +306,7 @@ func TestProxyURL_NoEnrollProxyAndProxyInThePolicy(t *testing.T) {
 
 func TestProxyURL_RemoveProxyFromThePolicy(t *testing.T) {
 	_ = define.Require(t, define.Requirements{
+		Group: "default",
```
Might be worth making the default group a constant in the testing package, mostly to prevent accidentally creating a different test group with a typo.

blakerouse replied:
Added.

cmacknz commented on testing/integration/upgrade_broken_package_test.go:
```diff
@@ -25,6 +25,7 @@ import (
 
 func TestUpgradeBrokenPackageVersion(t *testing.T) {
 	define.Require(t, define.Requirements{
+		Group: "upgrade-specific",
```
This is the only test group name that isn't self explanatory to me.

Maybe replace `upgrade-specific` with `upgrade-versions` since both this and `TestStandaloneDowngradeToSpecificSnapshotBuild` are about testing version related behavior?

blakerouse replied:
Okay.

cmacknz reviewed (commented):
Changes LGTM, a bunch of typos and some small clarifications to make before approval though.

michalpristas reviewed (approved):
aside from typos craig pointed out.

blakerouse commented:
We had some really old firewall rules that where created by OGC around in GCP. This was causing this to fail because it now creates more instances than it used to. OGC creates a firewall rule per instance and this was causing the quota to hit over 500 because there was around 400 old firewall rules around. I have cleaned those up, with hopes it allows the next run to pass.

blakerouse commented:
buildkite test this

blakerouse commented:
buildkite test this

pierrehilbert commented:
Bringing it back to draft as it has been open for a long time without any movement.

blakerouse commented:
I can see that the runner got stuck similar to the issue we have been seeing where it just stays running forever. With the groups it makes it easier to narrow it down to which group of tests is getting stuck:

I went through the logs and split out the required information to see what ran or still is running. Manually correlating each line to its runner. The breakdown is below and you can tell that `windows-amd64-2022-default` is stuck on `Running non-sudo tests...`

```
>>> (linux-arm64-ubuntu-2204-endpoint) Connected over SSH
>>> (linux-arm64-ubuntu-2204-endpoint) Running sudo tests...
>>> (linux-arm64-ubuntu-2204-endpoint) Test output (sudo) (stdout): >> go test: remote-linux-arm64-ubuntu-2204-endpoint-sudo.integration Test Passed

>>> (linux-arm64-ubuntu-2204-default) Connected over SSH
>>> (linux-arm64-ubuntu-2204-default) Running non-sudo tests...
>>> (linux-arm64-ubuntu-2204-default) Test output (non-sudo) (stdout): >> go test: remote-linux-arm64-ubuntu-2204-default.integration Test Passed
>>> (linux-arm64-ubuntu-2204-default) Running sudo tests...
>>> (linux-arm64-ubuntu-2204-default) Test output (sudo) (stdout): >> go test: remote-linux-arm64-ubuntu-2204-default-sudo.integration Test Passed

>>> (linux-arm64-ubuntu-2204-upgrade) Connected over SSH
>>> (linux-arm64-ubuntu-2204-upgrade) Running sudo tests...
>>> (linux-arm64-ubuntu-2204-upgrade) Test output (sudo) (stdout): >> go test: remote-linux-arm64-ubuntu-2204-upgrade-sudo.integration Test Passed

>>> (linux-arm64-ubuntu-2204-fleet-airgapped) Connected over SSH
>>> (linux-arm64-ubuntu-2204-fleet-airgapped) Running sudo tests...
>>> (linux-arm64-ubuntu-2204-fleet-airgapped) Test output (sudo) (stdout): >> go test: remote-linux-arm64-ubuntu-2204-fleet-airgapped-sudo.integration Test Passed

>>> (linux-arm64-ubuntu-2204-fleet) Connected over SSH
>>> (linux-arm64-ubuntu-2204-fleet) Running sudo tests...
>>> (linux-arm64-ubuntu-2204-fleet) Test output (sudo) (stdout): >> go test: remote-linux-arm64-ubuntu-2204-fleet-sudo.integration Test Passed

>>> (linux-amd64-ubuntu-2204-endpoint) Connected over SSH
>>> (linux-amd64-ubuntu-2204-endpoint) Running sudo tests...
>>> (linux-amd64-ubuntu-2204-endpoint) Test output (sudo) (stdout): >> go test: remote-linux-amd64-ubuntu-2204-endpoint-sudo.integration Test Passed

>>> (linux-amd64-ubuntu-2204-fleet) Connected over SSH
>>> (linux-amd64-ubuntu-2204-fleet) Running sudo tests...
>>> (linux-amd64-ubuntu-2204-fleet) Test output (sudo) (stdout): >> go test: remote-linux-amd64-ubuntu-2204-fleet-sudo.integration Test Passed

>>> (linux-amd64-ubuntu-2204-fleet-airgapped) Connected over SSH
>>> (linux-amd64-ubuntu-2204-fleet-airgapped) Running sudo tests...
>>> (linux-amd64-ubuntu-2204-fleet-airgapped) Test output (sudo) (stdout): >> go test: remote-linux-amd64-ubuntu-2204-fleet-airgapped-sudo.integration Test Passed

>>> (linux-amd64-ubuntu-2204-upgrade) Connected over SSH
>>> (linux-amd64-ubuntu-2204-upgrade) Running sudo tests...
>>> (linux-amd64-ubuntu-2204-upgrade) Test output (sudo) (stdout): >> go test: remote-linux-amd64-ubuntu-2204-upgrade-sudo.integration Test Passed

>>> (linux-amd64-ubuntu-2204-default) Connected over SSH
>>> (linux-amd64-ubuntu-2204-default) Running non-sudo tests...
>>> (linux-amd64-ubuntu-2204-default) Test output (non-sudo) (stdout): >> go test: remote-linux-amd64-ubuntu-2204-default.integration Test Passed
>>> (linux-amd64-ubuntu-2204-default) Running sudo tests...
>>> (linux-amd64-ubuntu-2204-default) Test output (sudo) (stdout): >> go test: remote-linux-amd64-ubuntu-2204-default-sudo.integration Test Passed

>>> (windows-amd64-2022-endpoint) Running sudo tests...
>>> (windows-amd64-2022-endpoint) Test output (sudo) (stdout): >> go test: remote-windows-amd64-2022-endpoint-sudo.integration Test Passed

>>> (windows-amd64-2022-upgrade) Running sudo tests...
>>> (windows-amd64-2022-upgrade) Test output (sudo) (stdout): >> go test: remote-windows-amd64-2022-upgrade-sudo.integration Test Passed

>>> (windows-amd64-2022-fleet) Running sudo tests...
>>> (windows-amd64-2022-fleet) Test output (sudo) (stdout): >> go test: remote-windows-amd64-2022-fleet-sudo.integration Test Passed

>>> (windows-amd64-2022-default) Connected over SSH
>>> (windows-amd64-2022-default) Running non-sudo tests...
```
